calicoctl/commands/ipam.go

```diff
@@ -0,0 +1,135 @@
+
+	docopt "github.com/docopt/docopt-go"
+)
+
+// IPAM show/release is for managing Calico IPAM by showing the
+// current use of an IP and/or releasing the IP address from use
+func IPAM(args []string) error {
+	doc := `Usage: 
+    calicoctl ipam release <IP>
+    calicoctl ipam show <IP>
+
+Description:
+    Manage Calico assigned IP addresses
+
+Warnings:
+  -  Releasing an in-use IP address can result in it being assigned to multiple
+     workloads.
+`
+
+	parsedArgs, err := docopt.Parse(doc, args, true, "calicoctl", false, false)
+	if err != nil {
+		return err
+	}
+
+	// length of parsedArgs is 0 when `-h` `--help` is passed
+	// docopts takes care of printing the help message, but we need to
+	// stop return nil here so command doesn't go any further when `-h` is passed
+	if len(parsedArgs) == 0 {
+		return nil
+	}
+
+	// Parse the input string as an IP address (IPv4 and IPv6)
+	// This also validates the IP address
+	ip := net.ParseIP(args[2])
+	if ip == nil {
+		fmt.Println("Invalid IP address specified.")
+		os.Exit(1)
+	}
+
+	// create a new backend client from env vars
+	backendClient, err := newClient("")
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	ipamClient := backendClient.IPAM()
+
+	switch args[1] {
+	case "show":
+		err = showIP(ipamClient, ip)
+	case "release":
+		err = releaseIP(ipamClient, ip)
+	}
+
+	// Exit with an error message if any of the command returned an error
+	if err != nil {
+		fmt.Printf("Error executing command %v, Error: %v\n", args, err)
+		os.Exit(1)
+	}
+
+	return nil
+}
+
+// showIP gets the attributes of an IP address, and returns nil if it is assigned
+// or an error with a message if not assigned
+func showIP(ipamClient client.IPAMInterface, ip net.IP) error {
+	attr, err := ipamClient.GetAssignmentAttributes(caliconet.IP{ip})
+
+	// IP address is not assigned, this prints message like
+	// `IP 192.168.71.1 is not assigned in block`. This is not exactly an error,
+	// so not returning it to the caller
+	if err != nil {
+		fmt.Println(err)
+		return nil
+	}
+
+	// IP address is assigned with attributes
+	if len(attr) != 0 {
+		fmt.Println(attr)
+	} else {
+		// IP address is assigned but attributes are not set
+		fmt.Printf("No attributes defined for IP %s\n", ip)
+	}
+
+	return nil
+}
+
+// releaseIP releases the IP address passed to it
+// or prints error message if it's already released
+func releaseIP(ipamClient client.IPAMInterface, ip net.IP) error {
+	ips := []caliconet.IP{caliconet.IP{ip}}
+
+	// Call ReleaseIPs releases the IP and returns an empty slice as unreleasedIPs if
+	// release was successful else it returns back the slice with the IP passed in
+	unreleasedIPs, err := ipamClient.ReleaseIPs(ips)
+	if err != nil {
+		fmt.Printf("Error: %v\n", err)
+		return err
+	}
+
+	// Couldn't release the IP if the slice is not empty
+	// or IP might already be released/unassigned. This is not exactly an error,
+	// so not returning it to the caller
+	if len(unreleasedIPs) != 0 {
```

Comment:
This case means that the given IP address wasn't assigned.  The message we print out should indicate that more clearly.

How about "IP address %s is not assigned" ?